runtime/src/accounts_db.rs

```diff
@@ -4364,6 +4365,23 @@ impl AccountsDb {
                             });
                         }
                     }
+                    if let Some(cache) = accounts_cache {
+                        if let Some(slot_cache) = cache.slot_cache(slot) {
```

Comment:
Can there be slots in the cache earlier than the `let start = snapshot_storages.range().start` we start checking from? https://github.com/solana-labs/solana/pull/17443/files#diff-1090394420d51617f3233275c2b65ed706b35b53b115fe65f82c682af8134a6fR4356

Reply:
Yes, it could break. This gets to another aspect of a 'min' slot concept I talked to you the other day about. An assumption now is that we have a storage for the 'min' slot we care about - it is either a root or now possibly an ancestor. For this pr's new usage of scanning stores, we also have to introduce the concept of a 'max_inclusive' slot to make sure we check the write cache for all slots we know about up to and including 'max_inclusive' because the slot range defined by the storages is not sufficient for all the slots we care about - specifically on the max side. The min side is the same issue. I guess if we used min(root, ancestor), we could bracket the other end. This specifically then gets to the topic of roots not knowing the min root in its current implementation (because of the potential for excess). We can fix this - there just wasn't a use case yet. Maybe this pr is a use case for the min. Basically, to scan stores and the cache, I need to know every slot in the range because there could be something in the write cache for that.  Well, actually, this is wrong as it is because I don't want EVERYTHING in the write cache - I only want slots that are a root or an ancestor. So, already this code is wrong... Ugh. The value of prs...  I imagined there was at least 1 thing this code wasn't solving yet...

Reply:
ok, I pushed a commit that pairs accounts cache with ancestors so we won't be adding account data for slots which were not our ancestors. I am assuming that a slot that is a root does not have data in the write cache? Is that true? @carllin

Reply:
> I am assuming that a slot that is a root does not have data in the write cache? Is that true?

Unfortunately this isn't true, the write cache can store many roots between flushes. On startup processing the cache is flushed every 10 seconds here https://github.com/solana-labs/solana/blob/bea7ce717c7bf68625cb9d5218d4460982cb3ef6/ledger/src/blockstore_processor.rs#L1025-L1031, which means there may be many roots present in the cache since the last flush

Reply:
Yeah it seems like the safest thing to do would either: 

1) Force flush of the cache before computing the capitalization to avoid this issue altogether
2) Start iteration from the smallest item in the cache to the first stored slot. Because at startup the cache only keeps the latest 10 seconds of roots around, the *earliest* root in the cache is  probably much greater than the *earliest* root tracked in the `RootsTracker`. Maybe we could track all the slots in the cache with something similar to a small `BTreeSet` and just iterate over that set to find any cache items less than the first stored slot?

Reply:
@carllin , please take another look. I did #2. I put in a few separate prs to take care of some simple code movement.